src/expr/src/relation/join_input_mapper.rs

```diff
@@ -354,6 +355,90 @@ impl JoinInputMapper {
         }
         None
     }
+
+    /// Try to find a consequence `c` of the given expression `e` for the given input.
+    ///
+    /// If we return `Some(c)`, that means
+    ///   1. `c` uses only columns from the given input;
+    ///   2. if `c` doesn't hold on a row of the input then `e` also wouldn't hold;
+    ///   3. if `c` holds on a row of the input then `e` might or might not hold.
+    /// 1. and 2. means that if we have a join with predicate `e` then we can use `c` for
+    /// pre-filtering a join input before the join. However, 3. means that `e` shouldn't be deleted
+    /// from the join predicates, i.e., we can't do a "traditional" predicate pushdown.
+    ///
+    /// Note that "`c` is a consequence of `e`" is the same thing as 2., see
+    /// <https://en.wikipedia.org/wiki/Contraposition>
+    ///
+    /// Example: For
+    /// `(t1.f2 = 3 AND t2.f2 = 4) OR (t1.f2 = 5 AND t2.f2 = 6)`
+    /// we find
+    /// `t1.f2 = 3 OR t1.f2 = 5` for t1, and
+    /// `t2.f2 = 4 OR t2.f2 = 6` for t2.
+    ///
+    /// Further examples are TPC-H Q07, Q19, and chbench Q07, Q19.
+    ///
+    /// Parameters:
+    ///  - `expr`: The expression `e` from above.
+    ///  - `index`: The index of the join input whose columns we will use.
+    ///  - `equivalences`: Join equivalences that we can use for `try_map_to_input_with_bound_expr`.
+    pub fn consequence_for_input(
+        &self,
+        expr: MirScalarExpr,
+        index: usize,
+        equivalences: &[Vec<MirScalarExpr>],
+    ) -> Option<MirScalarExpr> {
+        if let Some(localized) =
+            self.try_map_to_input_with_bound_expr(expr.clone(), index, equivalences)
+        {
+            Some(localized)
+        } else {
```

Comment:
To avoid calling `try_map_to_input_with_bound_expr` and `expr.clone()` multiple times, we can split `try_map_to_input_with_bound_expr` into two functions: 1) remaps the expression to refer to the given input whenever possible 2) checks that the expression has been entirely remapped to a given input. Then this method only needs to call the expression remapping method once, and the recursion only needs to check that the subexpression of the result all belongs to the given input.

Reply:
Good idea, thx!

Reply:
The `expr.clone()` was the only thing that bothered me at first glance. Great to see that there is a workaround for that!